bootstrap.go

```diff
@@ -1,57 +1,85 @@
-		logger.Errorf("Datadog configuration not valid, cannot initialize Datadog services")
-
-		return
-	}
-
-	logger.Infof("Initializing Datadog services for %s in environment %s", cfg.Service, cfg.Env)
-
-	tracer.Start(
-		tracer.WithEnv(cfg.Env),
-		tracer.WithUDS(cfg.APM),
-		tracer.WithService(cfg.Service),
-		tracer.WithServiceVersion(cfg.ServiceVersion),
-	)
-
-	var profileTypes []profiler.ProfileType
-	if enableExtraProfiling {
-		profileTypes = []profiler.ProfileType{
-			profiler.CPUProfile,
-			profiler.HeapProfile,
-			profiler.GoroutineProfile,
-			profiler.MutexProfile,
-			profiler.BlockProfile,
-		}
-	} else {
-		profileTypes = []profiler.ProfileType{profiler.CPUProfile}
-	}
-
-	err := profiler.Start(
-		profiler.WithEnv(cfg.Env),
-		profiler.WithUDS(cfg.APM),
-		profiler.WithService(cfg.Service),
-		profiler.WithVersion(cfg.ServiceVersion),
-		profiler.WithProfileTypes(profileTypes...),
-	)
-	if err != nil {
-		logger.Errorf("Failed to start Datadog profiler: %v", err)
-	}
+func StartDatadog(cfg config.DatadogParameters, enableExtraProfiling, isConnectionSocket bool) {
+    if !cfg.IsDataDogConfigValid() {
+        logger.Errorf("Datadog configuration not valid, cannot initialize Datadog services")
+
+        return
+    }
+
+    logger.Infof("Initializing Datadog services for %s in environment %s", cfg.GetService(), cfg.GetEnv())
+
+    initTracer(cfg, isConnectionSocket)
+    if initProfilerErr := initProfiler(cfg, enableExtraProfiling, isConnectionSocket); initProfilerErr != nil {
+        logger.Errorf("Failed to start Datadog profiler: %v", initProfilerErr)
+    }
 }
 
 // GracefulDatadogShutdown of executed parallel processes
 func GracefulDatadogShutdown() {
-	defer tracer.Stop()
-	defer profiler.Stop()
+    defer tracer.Stop()
+    defer profiler.Stop()
+}
+
+func initTracer(cfg config.DatadogParameters, isConnectionSocket bool) {
+    var tracerOptions []tracer.StartOption
+    if isConnectionSocket {
+        tracerOptions = append(tracerOptions, tracer.WithUDS(cfg.GetApmEndpoint()))
+    } else {
+        tracerOptions = append(tracerOptions, tracer.WithAgentAddr(cfg.GetApmEndpoint()))
+    }
+
+    tracerOptions = append(
+        tracerOptions,
+        []tracer.StartOption{
+            tracer.WithEnv(cfg.GetEnv()),
+            tracer.WithService(cfg.GetService()),
+            tracer.WithServiceVersion(cfg.GetServiceVersion()),
+        }...,
+    )
+
+    tracer.Start(tracerOptions...)
+}
+
+func initProfiler(cfg config.DatadogParameters, enableExtraProfiling bool, isConnectionSocket bool) error {
+    var profilerTypes []profiler.ProfileType
+    if enableExtraProfiling {
+        profilerTypes = []profiler.ProfileType{
+            profiler.CPUProfile,
+            profiler.HeapProfile,
+            profiler.GoroutineProfile,
+            profiler.MutexProfile,
+            profiler.BlockProfile,
+        }
+    } else {
+        profilerTypes = []profiler.ProfileType{profiler.CPUProfile}
+    }
```

Comment:
I think we should either allow the user of the library full control over the profiling or no control and always enable everything.

Alternatively we can default to everything unless the user provides a set of profiling profiles to be used.

Since this library is planed to go on the paved path good defaults are important.

Does anyone else have any input here @coopnorge/engineering

Reply:
I'm not sure I understand what you mean.

Reply:
I mean we can update this part of the code later when it will be needed to consumers configure the profiler. Right now most of the services either enable basic profiling for CPU or detailed for example in a staging environment.